### ISSUE TITLE
pyampute Type error, fit_transform attribute not found

### ISSUE THREAD

**Issue by prabhant** [RianneSchouten/pyampute#5]
Code

```
from pyampute.ampute import MultivariateAmputation
from sklearn.pipeline import make_pipeline
make_pipeline(MultivariateAmputation(),SimpleImputer())

```

Error
```
Traceback (most recent call last):
  File "idea.py", line 14, in <module>
    make_pipeline(MultivariateAmputation(),SimpleImputer())
  File "/Users/prabhant/opt/anaconda3/lib/python3.8/site-packages/sklearn/pipeline.py", line 715, in make_pipeline
    return Pipeline(_name_estimators(steps), memory=memory, verbose=verbose)
  File "/Users/prabhant/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/Users/prabhant/opt/anaconda3/lib/python3.8/site-packages/sklearn/pipeline.py", line 114, in __init__
    self._validate_steps()
  File "/Users/prabhant/opt/anaconda3/lib/python3.8/site-packages/sklearn/pipeline.py", line 159, in _validate_steps
    raise TypeError("All intermediate steps should be "
TypeError: All intermediate steps should be transformers and implement fit and transform or be the string 'passthrough' '<pyampute.ampute.MultivariateAmputation object at 0x7feaca220370>' (type <class 'pyampute.ampute.MultivariateAmputation'>) doesn't
```
This is very weird because
```
print(hasattr(MultivariateAmputation(), "fit_transform"))
outputs : True
```

Any idea why this is happening?

**Comment by prabhant** [comment]
@davzaman any idea about it?

**Comment by RianneSchouten** [comment]
@prabhant Can this have to do with line 860 in ampute.py? There, we print a missingness profile. 

Actually, I will make a todo that moves the missingness_profile to the mdPatterns class. I think it is more suitable there.

**Comment by davzaman** [comment]
This is because we only implement `fit_transform()` Pipeline requires that `fit()` and `transform()` are [implemented separately](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html)

**Comment by davzaman** [comment]
I can split it but it's not immediately obvious how we should split it. We want to fit and transform on the same data, so it becomes a little strange.

We can just let our `fit_transform()` be called `transform()`, and write a `fit()` function that does nothing so it doesn't complain, and then write a note to users that `fit()` does nothing, call `fit_transform()` or `transform()`. This is a bit weird.

Another way would be to save `X_incomplete`, `X_indices`, and `assigned_group_number` in `fit()`, and then apply them in `transform()`, and then ignore the input to `transform()`. This one is more in the spirit of "fit" then "transform" but looks very strange as a transformer. 

Thoughts?

**Comment by RianneSchouten** [comment]
Dear davina,

Thank you for finding the problem! I realize now that the idea of fit and
transform is that you can apply the same settings to different datasets
(such as train and test).

I like your last suggestion. Let's say that in fit we do all the checks and
we apply the defaults. Then in transform we divide cases over the patterns
and do the rest. I think the division of cases over the patterns is dataset
dependent and should therefore happen in transform instead of in fit, what
do you think?

In this way it should be possible for users to apply the same amputation
model to a train and test, or in a simulation to new version of the same
complete dataset, isn't it? What do you think?

**Comment by davzaman** [comment]
I think that it makes sense but there's some oddities that need ironing. For example, do we re-validate the dataset in transform?

We can possibly separate the validation for the patterns in fit vs validation for the dataset will happen in transform, and just use the data passed on fit to extract the number of features, but not actually do any dataset checks until transform. I think that should work.

**Comment by RianneSchouten** [comment]
Dear Davina,

I will have a closer look at it next week. It is difficult to fully
understand the consequences, and i therefore also tend to like your first
solution to make an 'empty' fit function.

Let me have a look next week and we can discuss on Monday.

Rianne

**Comment by prabhant** [comment]
sklearn code dictates we need fit/fit_transform and transform, 
The code works fine with an empty transform function

```
if not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not hasattr(
    t, "transform"
):
```

**Comment by prabhant** [comment]
this commit solves it https://github.com/RianneSchouten/pyampute/commit/1ec3abbaaa54cbf9fcaf722a27ad8c7a67c5059a

**Comment by davzaman** [comment]
@prabhant , thanks for taking a look. I think we want fit and transform to have meaningful logic (fit actually "fits" the amputer and transform just transforms, instead of having one be empty). @RianneSchouten and I discussed a bit further and came to a solution. It is fixed with eae5de3
